File: leetcode/weekly_420/4.py
from typing import List

class Solution:
    def findAnswer(self, parent: List[int], s: str) -> List[bool]:

        def is_palindrome(s: str):
            left = 0
            right = len(s) - 1
            while left < right:
                if s[left] != s[right]:
                    return False
                left += 1
                right -= 1
            return True

        edges = {}
        for i in range(1, len(parent)):
            if parent[i] not in edges:
                edges[parent[i]] = []
            edges[parent[i]].append(i)

        ans = [False for _ in range(len(parent))]

        def dfs(x: int) -> str:
            res = ""
            if x in edges:
                for dest in edges[x]:
                    res += dfs(dest)
            res += s[x]
            ans[x] = is_palindrome(res)
            return res

        dfs(0)
        return ans

File: leetcode/weekly_420/4.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:

	unordered_map<int, vector<int>> edges;

	bool is_palindrome(string& s) {
		int left = 0, right = s.length() - 1;
		while (left < right) {
			if (s[left] != s[right]) {
				return false;
			}
			left++;
			right--;
		}
		return true;
	}

	string dfs(int x, const string& s, vector<bool>& ans) {
		string res = "";
		for (int dest : edges[x]) {
			res += dfs(dest, s, ans);
		}

		res += s[x];
		ans[x] = is_palindrome(res);
		return res;
	}

    vector<bool> findAnswer(vector<int>& parent, string s) {
		for (int i = 1; i < parent.size(); i++) {
			edges[parent[i]].push_back(i);
		}

		vector<bool> ans(parent.size());
		dfs(0, s, ans);
		return ans;
    }
};
